detectable/src/main/java/com/synopsys/integration/detectable/detectable/Requirements.java

```diff
@@ -0,0 +1,153 @@
+import java.util.List;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import com.synopsys.integration.detectable.DetectableEnvironment;
+import com.synopsys.integration.detectable.ExecutableTarget;
+import com.synopsys.integration.detectable.detectable.exception.DetectableException;
+import com.synopsys.integration.detectable.detectable.explanation.Explanation;
+import com.synopsys.integration.detectable.detectable.explanation.FoundExecutable;
+import com.synopsys.integration.detectable.detectable.explanation.FoundFile;
+import com.synopsys.integration.detectable.detectable.file.FileFinder;
+import com.synopsys.integration.detectable.detectable.result.DetectableResult;
+import com.synopsys.integration.detectable.detectable.result.ExecutableNotFoundDetectableResult;
+import com.synopsys.integration.detectable.detectable.result.FileNotFoundDetectableResult;
+import com.synopsys.integration.detectable.detectable.result.PassedDetectableResult;
+
+public class Requirements {
+    private DetectableResult failure;
+    private List<Explanation> explanations = new ArrayList<>();
+    private List<File> relevantFiles = new ArrayList<>();
+
+    private final FileFinder fileFinder;
+    private final DetectableEnvironment environment;
+
+    public Requirements(final FileFinder fileFinder, final DetectableEnvironment environment) {
+        this.fileFinder = fileFinder;
+        this.environment = environment;
+    }
+
+    public void explain(Explanation explanation) {
+        explanations.add(explanation);
+    }
+
+    public void explainFile(@NotNull File file) {
+        relevantFiles.add(file);
+        explanations.add(new FoundFile(file));
+    }
+
+    public void explainNullableFile(@Nullable File file) {
+        if (file == null)
+            return;
+        explainFile(file);
+    }
+
+    public File file(final String filename) {
+        return file(environment.getDirectory(), filename);
+    }
+
+    public File optionalFile(final String filename, RequirementNotMetAction ifNotMet) {
+        return optionalFile(environment.getDirectory(), filename, ifNotMet);
+    }
+
+    public File optionalFile(File directory, final String filename, RequirementNotMetAction ifNotMet) {
+        if (isAlreadyFailed())
+            return null;
+
+        File file = fileFinder.findFile(directory, filename);
+        if (file == null) {
+            ifNotMet.requirementNotMet();
+        } else {
+            relevantFiles.add(file);
+            explanations.add(new FoundFile(file));
+        }
+        return file;
+    }
+
+    public File file(File directory, final String filename) {
+        //Only difference between Optional File and Required File is Required populate failure, so if optional 'is not met' we can capture that by setting failure.
+        return optionalFile(directory, filename, () -> {
+            failure = new FileNotFoundDetectableResult(filename);
+        });
```

Comment:
Seems like failure could get overwritten here if two files are required and both are missing? Would there be any value in making it a list, so in a case like that both failures are captured?

Reply:
The pattern we had was the first failure was returned so I've replicated that. In requirements case though each requirement will immediately return null if 'requiements' failed at large, so there is no worry about overwriting the failure right now. 

I'm guessing for Applicable we wouldn't see any benefits but for Extractable there might be, like if the executable and inspector both fail we can tell you now rather than fix one and then have to fix the other right away. We could make a composite failure result or modify detectable to return a list of results. I think thats a good idea for a future enhancement.